src/builtins/impls/time.rs

```diff
@@ -14,67 +14,147 @@
 
 /// Returns the time in nanoseconds parsed from the string in RFC3339 format.
 /// `undefined` if the result would be outside the valid time range that can fit
 /// within an `int64`.
 #[tracing::instrument(name = "time.parse_rfc3339_ns", err)]
 pub fn parse_rfc3339_ns(value: String) -> Result<i64> {
-    bail!("not implemented");
+    Ok(DateTime::parse_from_rfc3339(value.as_ref())?.timestamp_nanos())
 }
 
 /// Returns the day of the week (Monday, Tuesday, ...) for the nanoseconds since
 /// epoch.
 #[tracing::instrument(name = "time.weekday", err)]
 pub fn weekday(x: serde_json::Value) -> Result<String> {
-    bail!("not implemented");
+    let date_time = extract_data_from_value(&x)?;
+
+    Ok(match date_time.weekday() {
+        Weekday::Mon => "Monday",
+        Weekday::Tue => "Tuesday",
+        Weekday::Wed => "Wednesday",
+        Weekday::Thu => "Thursday",
+        Weekday::Fri => "Friday",
+        Weekday::Sat => "Saturday",
+        Weekday::Sun => "Sunday",
+    }
+    .to_string())
+}
+
+/// convert nanoseconds to [`chrono::DateTime`] by the given timezone
+///
+/// # Errors
+///
+/// - the `time_zone` when could not parse the given timezone
+/// - the given `ns` representation is invalid.
+fn nanoseconds_to_date(ns: i64, time_zone: &str) -> Result<DateTime<Tz>> {
+    let tz: Tz = match time_zone.parse() {
+        Ok(tz) => tz,
+        Err(e) => return Err(anyhow!("timezone parse error: {}", e)),
+    };
+
+    match tz.timestamp_opt(ns / 1_000_000_000, (ns % 1_000_000_000) as u32) {
+        LocalResult::None => return Err(anyhow!("No such local time")),
+        LocalResult::Single(t) => Ok(t),
+        LocalResult::Ambiguous(t1, t2) => {
+            return Err(anyhow!(
+                "ambiguous local time, ranging from {:?} to {:?}",
+                t1,
+                t2
+            ))
+        }
+    }
+}
+
+/// extract ns and timezone from `serde_json::Value`
+///
+/// expected from `value` to be an array of i64
+/// - when it array, take only the position 0 (ns) and position 1 (timezone) and
+///   convert to [`chrono::DateTime`]
+/// - when it i64, convert the value (with default timezone UTC) to a
+///   [`chrono::DateTime`]
+///
+/// # Errors
+///
+/// - the value is empty or has only one field
+/// - the single value is not i64
+/// - when has an error to convert ns to [`chrono::DateTime`]
+fn extract_data_from_value(value: &serde_json::Value) -> Result<DateTime<Tz>> {
+    let (ns, timezone) = match value.as_array() {
+        Some(_data) => match (value.get(0).unwrap_or(value).as_i64(), value.get(1)) {
+            (Some(ns), Some(timezone)) => (ns, timezone.as_str().unwrap_or_default().to_string()),
+            _ => return Err(anyhow!(ERROR_INVALID_ARGUMENTS)),
+        },
+        _ => match value.as_i64() {
+            Some(ns) => (ns, DEFAULT_TIMEZONE.to_string()),
+            None => return Err(anyhow!(ERROR_INVALID_ARGUMENTS)),
+        },
+    };
+    nanoseconds_to_date(ns, timezone.as_str())
```

Comment:
If I used `serde_json::Value` as builtin arguments is that I was lazy and did not defined a custom type.

What you can do instead is define a type like this:

```rust
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum TimestampWithOptionalTimezone {
    Timestamp(i64),
    TimestampAndTimezone(i64, String),
}
```

Reply:
you mean that the function `nanoseconds_to_date` will receive `TimestampWithOptionalTimezone` enum
```rs
fn nanoseconds_to_date(timestamp: TimestampWithOptionalTimezone) -> Result<DateTime<Tz>> {
...
}
```

And replace `let (ns, timezone) = match value.as_array() {...` to  `let timestamp: TimestampWithOptionalTimezone = match value.as_array() {...`

Reply:
Not exactly: builtins are implemented in a way where parameters can be anything that can be deserialised by serde.

This means you could define this `TimestampWithOptionalTimezone` and use it directly in the builtins parameters like

```
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum TimestampWithOptionalTimezone {
    Timestamp(i64),
    TimestampAndTimezone(i64, String),
}

impl TimestampWithOptionalTimezone {
    fn to_datetime(self) -> Result<Datetime<Tz>> {
        match self {
            Timestamp(ts) => todo!(),
            TimestampAndTimezone(ts, tz) => todo!(),
        }
    }
} 

pub fn weekday(x: TimestampWithOptionalTimezone) -> Result<&'static str> {
    let date_time = x.to_datetime()?;
    Ok(match date_time.weekday() {
        Weekday::Mon => "Monday",
        Weekday::Tue => "Tuesday",
        Weekday::Wed => "Wednesday",
        Weekday::Thu => "Thursday",
        Weekday::Fri => "Friday",
        Weekday::Sat => "Saturday",
        Weekday::Sun => "Sunday",
    })
}